common/src/main/java/org/conscrypt/ClientSessionContext.java

```diff
@@ -131,6 +147,42 @@ private NativeSslSession getSession(String host, int port) {
         return null;
     }
 
+    private void putSession(HostAndPort key, NativeSslSession session) {
+        synchronized (sessionsByHostAndPort) {
+            List<NativeSslSession> sessions = sessionsByHostAndPort.get(key);
+            if (sessions == null) {
+                sessions = new ArrayList<NativeSslSession>();
+                sessionsByHostAndPort.put(key, sessions);
+            }
+            // To maintain the invariant that single- and multi-use sessions aren't
+            // mixed, check what the current list contains and remove those sessions if
+            // they're of the other type.
+            if (sessions.size() > 0) {
+                if (sessions.get(0).isSingleUse() != session.isSingleUse()) {
+                    while (!sessions.isEmpty()) {
+                        removeSession(sessions.get(0));
+                    }
+                    // The last removeSession() call will have removed the list from
+                    // the map, so put it back.
+                    sessionsByHostAndPort.put(key, sessions);
+                }
+            }
+            sessions.add(session);
+        }
+    }
+
+    private void removeSession(HostAndPort key, NativeSslSession session) {
```

Comment:
I notice you have two different `removeSession` methods here, one which takes a key and the other one on the abstract class which assumes the key is the session ID. Perhaps a different name?